libfreerdp/core/proxy.c

```diff
@@ -255,3 +284,181 @@ BOOL http_proxy_connect(BIO* bufferedBio, const char* hostname, UINT16 port)
 	return TRUE;
 }
 
+static int recv_socks_reply(BIO* bufferedBio, BYTE* buf, int len, char* reason, BYTE checkVer)
+{
+	int status;
+
+	for(;;)
+	{
+		status = BIO_read(bufferedBio, buf, len);
+		if (status > 0)
+			break;
+
+		if (status < 0)
+		{
+			/* Error? */
+			if (BIO_should_retry(bufferedBio))
+			{
+				USleep(100);
+				continue;
+			}
+
+			WLog_ERR(TAG, "Failed reading %s reply from SOCKS proxy (Status %d)", reason, status);
+			return -1;
+		}
+
+		if (status == 0)
+		{
+			/* Error? */
+			WLog_ERR(TAG, "Failed reading %s reply from SOCKS proxy (BIO_read returned zero)", reason);
+			return -1;
+		}
+	}
+	
+	if (status < 2)
+	{
+		WLog_ERR(TAG, "SOCKS Proxy reply packet too short (%s)", reason);
+		return -1;
+	}
+
+	if (buf[0] != checkVer)
+	{
+		WLog_ERR(TAG, "SOCKS Proxy version is not 5 (%s)", reason);
+		return -1;
+	}
+
+	return status;
+}
+
+/* CONN REQ replies in enum. order */
+static const char *rplstat[] =
+{
+	"succeeded",
+	"general SOCKS server failure",
+	"connection not allowed by ruleset",
+	"Network unreachable",
+	"Host unreachable",
+	"Connection refused",
+	"TTL expired",
+	"Command not supported",
+	"Address type not supported"
+};
+
+
+static BOOL socks_proxy_connect(BIO* bufferedBio, const char *proxyUsername, const char *proxyPassword,
+		const char* hostname, UINT16 port)
+{
+	int status;
+	int nauthMethods = 1, writeLen = 3;
+	BYTE buf[3 + 255 + 255]; /* biggest packet is user/pass auth */
+	size_t hostnlen = strnlen(hostname, 255);
+
+	if (proxyUsername && proxyPassword)
+	{
+		nauthMethods++;
+		writeLen++;
+	}
+
+	/* select auth. method */
+	buf[0] = 5; /* SOCKS version */
+	buf[1] = nauthMethods; /* #of methods offered */
+	buf[2] = AUTH_M_NO_AUTH;
+	if (nauthMethods > 1)
+		buf[3] = AUTH_M_USR_PASS;
+
+	status = BIO_write(bufferedBio, buf, writeLen);
+	if (status != writeLen)
+	{
+		WLog_ERR(TAG, "SOCKS proxy: failed to write AUTH METHOD request");
+		return FALSE;
+	}
+
+	status = recv_socks_reply(bufferedBio, buf, 2, "AUTH REQ", 5);
+	if (status <= 0)
```

Comment:
Should be `<= 1`

Reply:
I've added a check recv_socks_reply so that status is at least 2 when we read some data.